File: backend/app/schemas/admin/knowledge_base.py
#!/usr/bin/python
# -*- coding:utf-8 -*-
# @author  : Liu Lijun
# @date    : 2026-06-13
# @description: Pydantic schemas for KnowledgeBase CRUD

from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Shared helpers
# ---------------------------------------------------------------------------

class _KnowledgeBaseBase(BaseModel):
    """Fields shared by Create / Update requests."""

    name: str = Field(..., max_length=100, description="Knowledge base name")    
    domain_id: int = Field(..., description="Domain this KB belongs to")
    keywords: Optional[list[str]] = Field(None, description="List of keywords for SmartRouter keyword matching")
    description: Optional[str] = Field(None, description="Knowledge base description")
    collection_name: str = Field(..., max_length=100, description="VectorDB collection name")
    embedding_id: Optional[int] = Field(None, description="Embedding configuration ID")
    chunk_size: int = Field(400, description="Block size (number of characters)")
    chunk_overlap: int = Field(80, description="Block overlap (number of characters)")
    parent_size: int = Field(1800, description="Parent block size (number of characters)")
    parent_overlap: int = Field(200, description="Parent block overlap (number of characters)")

    llm_id: Optional[int] = Field(None, description="LLM configuration ID")
    is_active: bool = Field(True, description="Activate or not?")

# ---------------------------------------------------------------------------
# Request schemas
# ---------------------------------------------------------------------------

class KnowledgeBaseCreate(_KnowledgeBaseBase):
    """Body for POST /knowledge_bases — all core fields are required."""

    name: str = Field(..., max_length=100, description="Knowledge base name")
    domain_id: int = Field(..., description="Domain this KB belongs to")
    collection_name: str = Field(..., max_length=100, description="VectorDB collection name")


class KnowledgeBaseUpdate(_KnowledgeBaseBase):
    """Body for PATCH /knowledge_bases/{id} — all fields optional (partial update)."""

    pass


# ---------------------------------------------------------------------------
# Response schemas
# ---------------------------------------------------------------------------

class KnowledgeBaseRead(_KnowledgeBaseBase):
    """Full knowledge base representation returned to the client."""

    id: int
    document_count: int = Field(0, description="Number of documents")
    chunk_count: int = Field(0, description="Number of blocks")
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class KnowledgeBaseListOut(BaseModel):
    """Paginated list wrapper."""

    total: int
    page: int
    page_size: int
    items: list[KnowledgeBaseRead]


class KnowledgeBaseOption(BaseModel):
    """Knowledge base option for dropdown selection."""

    id: int
    name: str

    model_config = {"from_attributes": True}


class KnowledgeBaseStats(BaseModel):
    """Knowledge base statistics."""

    id: int
    name: str
    document_count: int
    chunk_count: int
    created_at: datetime
    updated_at: datetime
    is_active: bool

File: management/src/api/knowledge_base.ts
import { http } from "@/utils/http";
import { baseUrlApi } from "@/api/utils";

// ─── Types ────────────────────────────────────────────────────────────────────

export interface KnowledgeBaseItem {
  id: number;
  name: string;
  domain_id: number;
  keywords: string[] | null;
  description: string | null;
  collection_name: string;
  embedding_id: number | null;
  chunk_size: number;
  chunk_overlap: number;
  parent_size: number;
  parent_overlap: number;
  llm_id: number | null;
  is_active: boolean;
  document_count: number;
  chunk_count: number;
  created_at: string;
  updated_at: string;
}

export interface KnowledgeBaseListResult {
  total: number;
  page: number;
  page_size: number;
  items: KnowledgeBaseItem[];
}

export interface KnowledgeBaseStats {
  id: number;
  name: string;
  document_count: number;
  chunk_count: number;
  created_at: string;
  updated_at: string;
  is_active: boolean;
}

export interface KnowledgeBaseOption {
  id: number;
  name: string;
}

export interface KbListQuery {
  name?: string;
  domain_id?: number;
  is_active?: boolean;
  page: number;
  page_size: number;
}

export interface KbCreatePayload {
  name: string;
  domain_id: number;
  collection_name: string;
  keywords?: string[];
  description?: string;
  embedding_id?: number;
  chunk_size?: number;
  chunk_overlap?: number;
  parent_size?: number;
  parent_overlap?: number;
  llm_id?: number;
  is_active?: boolean;
}

export type KbUpdatePayload = Partial<KbCreatePayload>;

// ─── API calls ────────────────────────────────────────────────────────────────

/** GET /knowledge_bases  — paginated list */
export const getKnowledgeBaseList = (params: KbListQuery) =>
  http.request<KnowledgeBaseListResult>(
    "get",
    baseUrlApi("admin/knowledge-bases"),
    {
      params
  });

/** GET /knowledge_bases/options */
export const getKnowledgeBaseOptions = () =>
  http.request<{ data: KnowledgeBaseOption[] }>(
    "get",
    baseUrlApi("admin/knowledge-bases/options")
  );

/** GET /knowledge_bases/:id */
export const getKnowledgeBase = (id: number) =>
  http.request<KnowledgeBaseItem>(
    "get",
    baseUrlApi(`admin/knowledge-bases/${id}`)
  );

/** GET /knowledge_bases/:id/stats */
export const getKnowledgeBaseStats = (id: number) =>
  http.request<KnowledgeBaseStats>(
    "get",
    baseUrlApi(`admin/knowledge-bases/${id}/stats`)
  );

/** POST /knowledge_bases */
export const createKnowledgeBase = (data: KbCreatePayload) =>
  http.request<KnowledgeBaseItem>("post", baseUrlApi("admin/knowledge-bases"), {
    data
  });

/** PATCH /knowledge_bases/:id */
export const updateKnowledgeBase = (id: number, data: KbUpdatePayload) =>
  http.request<KnowledgeBaseItem>(
    "patch",
    baseUrlApi(`admin/knowledge-bases/${id}`),
    { data }
  );

/** DELETE /knowledge_bases/:id */
export const deleteKnowledgeBase = (id: number) =>
  http.request("delete", baseUrlApi(`admin/knowledge-bases/${id}`));

/** PATCH /knowledge_bases/:id/toggle */
export const toggleKnowledgeBase = (id: number) =>
  http.request("patch", baseUrlApi(`admin/knowledge-bases/${id}/toggle`));

export const getDomainOptions = () => {
  return http.request<{ id: number; name: string }[]>(
    "get",
    baseUrlApi("admin/domains/options")
  );
};

export const getEmbeddingOptions = () => {
  return http.request<{ id: number; name: string }[]>(
    "get",
    baseUrlApi("admin/embeddings/options")
  );
};

export const getLlmOptions = () => {
  return http.request<{ id: number; name: string }[]>(
    "get",
    baseUrlApi("admin/llms/options")
  );
};
